### PR TITLE
Add Proposal Derestricted Operators.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**

### CODE REVIEW

File: proposals/proposal-DerestrictedOperators.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+
+
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+## Alternatives
```
Comment (sharwell) [inline proposals/proposal-DerestrictedOperators.md]: 💭 In your specific example, you could apply `ObsoleteAttribute` to the operator which shouldn't be used, and set the `error` parameter to true.

**Comment by AdamSpeight2008** [comment]
@sharwell But that suggest that operator was supported (at some time) and is now obsolete, this is not true. Where as operator '>' not defined doesn't.  I see no reason to require the complement pairs, when the exception is throw anyways. Existing types are still supported as they've have implementations of those operators.

**Comment by sharwell** [comment]
@AdamSpeight2008 I mentioned it because it achieves nearly¹ the same final outcome without changing the language. In other words, it's an alternative whether or not it's a *good* alternative.

In my local test, I set the message for ObsoleteAttribute to "Operator > has no meaning for type X", which would be hard to mistake as meaning "Operator > used to be supported here".

¹ No warning would be reported if you used the operator in other code which is marked obsolete. Also it doesn't eliminate the need to actually define such a member.

**Comment by AdamSpeight2008** [comment]
The proposed change is to make it not an `error` but `warning` or lower.

File: proposals/proposal-DerestrictedOperators.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+`Error	BC30452	Operator '>' is not defined for types 'CMPOP_A(Of Integer)' and 'Integer'.`
+
+
+
+
+
+
+## Drawbacks
```
Comment (sharwell) [inline proposals/proposal-DerestrictedOperators.md]: The biggest drawback I see was actually caused by a misinterpretation. When I read this:

> This proposal is to remove that restriction, by skip the following section of under a language feature conditional.

It was my expectation that I was about to read about how defining `operator ==` alone would result in the compiler implementing `operator !=` as `!(x == y)`, and implementing `operator <` and `operator <=` could allow the compiler to implement the remaining two operators. The proposal here would preclude the ability to implement such a feature in a future release.

Comment (AdamSpeight2008) [inline proposals/proposal-DerestrictedOperators.md]: You're assuming that the developer intends them to mean "Comparison Operators".
How does it prevent a codefix from implement the complement for you?

Comment (sharwell) [inline proposals/proposal-DerestrictedOperators.md]: > How does it prevent a codefix from implement the complement for you?

I wasn't talking about a code fix. I thought you were proposing that if you included `==` and left out `!=`, the compiled output would still have an operator `!=` that was created behind the scenes. Similar to how if you create a class without a constructor then a default constructor is still created by the compiler.

**Comment by franzalex** [comment]
Unless I'm mistaken, these operators being referenced here are indeed comparison operators. For that matter, any given pair are expected to be mutually exclusive and defining one without the other can be expected to lead to a lacuna which will possibly lead to bugs later on. It will be rather unusual that an object can test for equality but not inequality, or greater but not less or equal. In a mutually exclusive case, the moment one condition is satisfied, it implicitly means that the reverse becomes false.

Granted that this restriction is removed as suggested by @AdamSpeight2008, it will be imperative that the suggestion proposed by @sharwell (automatic compiler generation of the operator pair) be implemented at the same time in order to 'save us from ourselves'.

**Comment by AdamSpeight2008** [comment]
It could be offered as suggestion, if you define a binary operator that returns a boolean.
Eg `Operator < ( L As T0, R As T0 ) As Boolean`
Which may suggest that they're are implementing a comparison operator.
That way it still permits non-compliment operator to be defined.

**Comment by AdamSpeight2008** [comment]
Implement a version that supports it.
vb noncomplementpairoperator
We could have the code suggestion, hook on that information message.

Just spotted that the compiler doesn't require complement pairs;=

|   Op   | With | 
| :----: |:----:|
|   =    |  <>  |
|   <    |  =>  |
|   <=   |  >   |

but mirrored pairs.

|   Op   | With | 
| :----: |:----:|
|   =    |  <>  |
|   <    |  >   |
|   <=   |  =>  |

Which is kinda surprising.

**Pull request merged**